common/expression/src/kernels/take.rs

```diff
@@ -0,0 +1,125 @@
+use crate::types::StringType;
+use crate::types::ValueType;
+use crate::with_number_type;
+use crate::Chunk;
+use crate::Column;
+use crate::Value;
+
+impl Chunk {
+    pub fn take<I: Index>(self, indices: &[I]) -> Result<Self> {
+        if indices.is_empty() {
+            return Ok(self.slice(0..0));
+        }
+
+        let mut after_columns = Vec::with_capacity(self.num_columns());
+        for value in self.columns() {
+            match value {
+                Value::Scalar(v) => after_columns.push(Value::Scalar(v.clone())),
+                Value::Column(c) => after_columns.push(Value::Column(Column::take(c, indices))),
+            }
+        }
+        Ok(Chunk::new(after_columns, indices.len()))
+    }
+}
+
+impl Column {
+    pub fn take<I: Index>(&self, indices: &[I]) -> Self {
+        let length = indices.len();
+        with_number_type!(SRC_TYPE, match self {
+            Column::SRC_TYPE(values) => {
+                Column::SRC_TYPE(Self::take_primitives(values, indices))
+            }
+            Column::Null { .. } | Column::EmptyArray { .. } => self.slice(0..length),
+
+            Column::Boolean(bm) => Self::take_scalars::<BooleanType, _>(
+                bm,
+                MutableBitmap::with_capacity(length),
+                indices
+            ),
+            Column::String(column) => Self::take_scalars::<StringType, _>(
+                column,
+                StringColumnBuilder::with_capacity(length, 0),
+                indices
+            ),
+            Column::Array(column) => {
+                let mut builder = ArrayColumnBuilder::<AnyType>::from_column(column.slice(0..0));
+                builder.reserve(length);
+                Self::take_scalars::<ArrayType<AnyType>, _>(column, builder, indices)
+            }
+            Column::Nullable(c) => {
+                let column = c.column.take(indices);
+                let validity = Self::take_scalars::<BooleanType, _>(
+                    &c.validity,
+                    MutableBitmap::with_capacity(length),
+                    indices,
+                );
+                Column::Nullable(Box::new(NullableColumn {
+                    column,
+                    validity: BooleanType::try_downcast_column(&validity).unwrap(),
+                }))
+            }
+            Column::Tuple { fields, .. } => {
+                let fields = fields.iter().map(|c| c.take(indices)).collect();
+                Column::Tuple {
+                    fields,
+                    len: indices.len(),
+                }
+            }
+        })
+    }
+
+    fn take_scalars<T: ValueType, I: Index>(
+        col: &T::Column,
+        mut builder: T::ColumnBuilder,
+        indices: &[I],
+    ) -> Column {
+        unsafe {
+            for index in indices {
+                T::push_item(
+                    &mut builder,
+                    T::index_column_unchecked(col, index.to_usize()),
+                )
+            }
+        }
```

Comment:
```suggestion
        let col = T::column_from_iter(indices.map(|index| unsafe { T::index_column_unchecked(col, index.to_usize()) }), &[]);
```

Reply:
`column_from_iter` is from trait `ArgType` but `AnyType` did not implement it.

I will create ` take_arg_types<T: ArgType>` and ` take_value_types<T: ValueType>` to dispatch them.

Reply:
maybe we can move it to,ValueType